slatedb/src/compactor_state.rs

```diff
@@ -260,6 +260,72 @@ impl CompactorState {
             last_sr_id = sr.id;
         }
     }
+
+    fn validate_compaction(&mut self, compaction: &Compaction) -> Result<(), SlateDBError> {
+        // Logical order of sources: [L0 (newest → oldest), then SRs (highest id → 0)]
+        let sources_logical_order: Vec<SourceId> = self
+            .db_state()
+            .l0
+            .iter()
+            .map(|sst| SourceId::Sst(sst.id.unwrap_compacted_id()))
+            .chain(
+                self.db_state()
+                    .compacted
+                    .iter()
+                    .map(|sr| SourceId::SortedRun(sr.id)),
+            )
+            .collect();
+
+        // Validate compaction sources exist
+        if compaction.sources.is_empty() {
+            warn!("submitted compaction is empty: {:?}", compaction.sources);
+            return Err(SlateDBError::InvalidCompaction);
+        }
+
+        // Validate if the compaction sources are strictly consecutive elements in the db_state sources
+        if !sources_logical_order
+            .windows(compaction.sources.len())
+            .any(|w| w == compaction.sources.as_slice())
+        {
+            warn!("submitted compaction is not a consecutive series of sources from db state: {:?} {:?}",
+            compaction.sources, sources_logical_order);
+            return Err(SlateDBError::InvalidCompaction);
+        }
+
+        let has_sr = compaction
+            .sources
+            .iter()
+            .any(|s| matches!(s, SourceId::SortedRun(_)));
+
+        if has_sr {
+            // Must merge into the lowest-id SR among sources
+            let min_sr = compaction
+                .sources
+                .iter()
+                .filter_map(|s| s.maybe_unwrap_sorted_run())
+                .min()
+                .expect("at least one SR in sources");
+            if compaction.destination != min_sr {
+                warn!(
+                    "destination does not match lowest-id SR among sources: {:?} {:?}",
+                    compaction.destination, min_sr
+                );
+                return Err(SlateDBError::InvalidCompaction);
+            }
+        } else {
+            // L0-only: must create new SR with id = highest_existing + 1
+            let next_id = self.db_state().compacted.first().map_or(0, |sr| sr.id + 1);
+            if compaction.destination != next_id {
```

Comment:
this can just check that `compaction.destination > compacted.first().unwrap_or(0)`. It doesn't need to be exactly 1 higher.

Reply:
That's just a detail of the scheduler. The requirement we need to check is a little looser.